tools/lsp/preview/drop_location.rs

```diff
@@ -742,12 +742,76 @@ fn find_filtered_location(
 }
 
 /// Find the Element to insert into. None means we can not insert at this point.
-pub fn can_drop_at(position: LogicalPoint, component_type: &str) -> bool {
-    let dm = &preview::component_instance()
-        .and_then(|ci| find_drop_location(&ci, position, component_type));
+pub fn can_drop_at(
+    document_cache: &common::DocumentCache,
+    position: LogicalPoint,
+    component: &common::ComponentInformation,
+) -> bool {
+    // let dm = &preview::component_instance()
+    //     .and_then(|ci| find_drop_location(&ci, position, component_type));
+
+    // preview::set_drop_mark(&dm.as_ref().and_then(|dm| dm.drop_mark.clone()));
+    // dm.is_some()
+
+    let Some(component_instance) = preview::component_instance() else {
+        return false;
+    };
+
+    let dm = find_drop_location(&component_instance, position, &component.name);
+
+    let can_drop = if let Some(dm) = &dm {
+        // Cache compilation results:
+        #[derive(Clone, Debug, Hash, Eq, PartialEq)]
+        struct CacheEntry {
+            component_type: String,
+            target_element: by_address::ByAddress<object_tree::ElementRc>,
+            target_node_index: usize,
+            child_index: usize,
+        }
+        static mut CACHE: std::cell::OnceCell<RefCell<clru::CLruCache<CacheEntry, bool>>> =
+            std::cell::OnceCell::new();
+
+        // SAFETY: This uses the document_cache, which means it runs in the UI thread.
+        let cache = unsafe {
```

Comment:
Ouch!  this should have been a thread_storage.
(and this produces warning in nightly rust even)